storage/heap/hp_write.c

```diff
@@ -45,8 +45,8 @@ int heap_write(HP_INFO *info, const uchar *record)
 #endif
 
   if ((share->records >= share->max_records && share->max_records) ||
-      (share->recordspace.total_data_length + share->index_length >=
-       share->max_table_size))
+      ((share->recordspace.total_data_length + share->index_length >=
+       share->max_table_size) && share->recordspace.del_chunk_count == 0)  )
```

Comment:
Formatting: `)  )`

Reply:
It looks like `share->recordspace.del_chunk_count == 0` condition is not 100% correct. In general we may be requesting more than 1 chunk, so in my opinion, the correct one is `share->recordspace.del_chunk_count < chunk_count` but it should be moved after `chunk_count` is calculated in `hp_get_encoded_data_length(share, record, &chunk_count);`
So, something like this
```c++
  if ((share->records >= share->max_records && share->max_records))
  {
    my_errno= HA_ERR_RECORD_FILE_FULL;
    DBUG_RETURN(my_errno);
  }

  hp_get_encoded_data_length(share, record, &chunk_count);

  if ((share->recordspace.total_data_length + share->index_length >=
       share->max_table_size) && share->recordspace.del_chunk_count < chunk_count)  )
  {
    my_errno= HA_ERR_RECORD_FILE_FULL;
    DBUG_RETURN(my_errno);
  }
  ...
```

Reply:
@percona-ysorokin , Current implementation handles this case properly. Allocation of more than 1 chunk is possible if the record is not fixed size. Even if we have only 1 chunk free, but record needs 2 chunks the situation will be handled in hp_allocate_chunkset() and we will return with error on next 'if'.
However I see that hp_allocate_chunkset() does not set my_errno, so we will get some random error => your proposal improves this, so let's go this way.